<?xml version="1.0" encoding="UTF-8" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
    <head>
        <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
        <title>topos.org.uk - Using QuickCheck to generate test datasets.</title>
        <link rel="stylesheet" type="text/css" href="../css/default.css" />
        <link rel="stylesheet" type="text/css" href="../css/syntax.css" />
    </head>
    <body>
        <div id="header">
            <div id="logo">
                <a href="../">topos.org.uk</a>
            </div>
            <div id="navigation">
                <a href="../">Home</a>
                <a href="../about.html">About</a>
                <a href="../contact.html">Contact</a>
                <a href="../archive.html">Archive</a>
            </div>
        </div>

        <div id="content">
            <h1>Using QuickCheck to generate test datasets.</h1>

            <div class="info">
    Posted on October 14, 2013
    
        by Nick
    
</div>

<p>Recently, I’ve been working on implementing annotations for the <a href="https://github.com/samtools/samtools">Samtools</a> library for dealing with aligned genetic sequence data. An ‘annotation’ in this terminology is a feature associated with some part of the file, be it a specific site in the genome, a specific phase of the underlying read (when the sequence was read from DNA) or the entire file. In this particular case, I’ve been looking at annotations associated with the entire file; features calculated by looking at all the sequence data which we then intend to feed into a regression system to tell us how ‘good’ we think the particular sample is.</p>
<p>One of the tests we’d like to perform is to determine whether there’s any bias in the positions along the genome of reads that are mapped as ‘variants’ - i.e. they do not correspond to the corresponding allele in the reference genome. On a large scale, there should be no correlation between the position of the read in the genome and its propensity to be a variant (this is not true on a small scale - there exist regions where variants are far more likely, but the location of these regions is roughly uniform). This sort of test - trying to determine whether the positions of two sets of data are different - is an obvious candidate for the <a href="http://en.wikipedia.org/wiki/Mann–Whitney_U">Mann-Whitney U Test</a>. The Mann-Whitney U (or Wilcoxon Rank-Sum - the two tests are subtly different but trivially related) is a non-parametric test of the null hypothesis that two groups of observations are drawn from the same distribution. It’s non-parametric because it doesn’t assume normality of the two underlying distributions (very useful in this case, where that’s an explicit non-assumption), and it works only by considering the ranks of items in each group in an ordering of the two groups.</p>
<p>This is particularly useful in this scenario - because we are comparing the positions of reads, we can calculate the Mann-Whitney U in a single pass through the data with no need to perform an ordering.</p>
<p>Samtools is written in C, but a recent venture with the language has convinced me that it’s nowhere I want to voluntarily be, so I’ve been writing code to test whether this is a useful annotation in Haskell. One of the advantages of this is the wonderful <a href="http://en.wikipedia.org/wiki/QuickCheck">QuickCheck</a> library. For those of you who haven’t come across it, the premise is that you define some sort of invariant you wish your code to satisfy and the library generates random data suitable to test that invariant. In this case, we were interested in making sure that the Normal approximation to the U distribution was suitably accurate.</p>
<p>QuickCheck works through the <code>Arbitrary</code> typeclass. An instance for <code>Arbitrary a</code> gives a mechanism to generate an arbitrary instance of <code>a</code>. In this case, we were looking to test that for arbitrary pairs of lists of integers, given they were of suitable size, the normal approximation to the U-distribution is sufficiently close to the exact calculation.</p>
<p>An advantage of QuickCheck which I didn’t really take advantage of here is that, once you have an <code>Arbitrary</code> instance for <code>a</code>, it’s trivial to define one for <code>[a]</code> or <code>(a, a)</code>. In this case, I wanted to enforce additional constraints (and it was my first time using quickcheck), so I wrote one directly:</p>
<pre><code>newtype TestRange = TestRange { unTestRange :: ([Int], [Int]) } deriving (Eq, Ord, Show)
instance Arbitrary TestRange where
  arbitrary = do
    n1 &lt;- choose (1,32) :: Gen Int -- size of the list
    n2 &lt;- choose (1,32) :: Gen Int -- size of second list
    randoms &lt;- forM [0 .. (n1 + n2)] (\_ -&gt; choose (1,20) :: Gen Int)
    return $ TestRange $ splitAt n1 $ randoms</code></pre>
<p>With this in hand, it’s very easy to write quick tests of invariants:</p>
<pre><code>prop_normalApproxEqualsExactRegion :: TestRange -&gt; Property
prop_normalApproxEqualsExactRegion (TestRange (a,b)) = let mag = length a * length b in
(mag &gt; 64 &amp;&amp; mag &lt; 100) ==&gt; abs (mwNormal a b - mwExact a b) &lt; 0.05</code></pre>
<p>This test was confirming that using the product of the sizes of the two groups was a suitable threshold to decide when to approximate. The usual literature suggests that the normal approximation is ‘good’ for sample sizes bigger than 8, but mostly considers sample sizes which are roughly equal. We expect very unequal sample sizes, and so may frequently have situations where one group is of size 2 and another may be in the thousands, making calculation of the exact U statistic unfeasible. Using a region bound (which seems just as good an indicator of test performance) makes these situations much easier to handle.</p>
<p>After mentioning that I was testing on random data, my collaborator asked me to produce some charts showing the performance of the approximation with the product of the group sizes, as well as exploring the relative performance of using the standard normal approximation, and an approximation which corrects for the presence of tied ranks in the input data. Luckily, it’s quite easy to use the arbitrary instances directly to generate data for export or graphing.</p>
<p>We define the type of test result:</p>
<pre><code>-- | TestResult n1 n2 exact normal uncorrected
data TestResult = TestResult Int Int Double Double Double deriving Show</code></pre>
<p>A test result contains the size of the two groups, as well as scores for the exact U-distribution, the normal approximation, and the normal approximation without correction for ties.</p>
<p>We could define an Arbitrary instance for <code>TestResult</code>, but in this case we want to make the call a little more explicit, so we simply expose the generator of <code>TestResults</code> directly:</p>
<pre><code>genResult :: Gen TestResult
genResult = let 
    filter (TestRange (a,b)) = (length a) * (length b) &lt; 200
  in do
    TestRange (a,b) &lt;- suchThat (arbitrary :: Gen TestRange) filter
    let u@(MWData _ n m _ _ ) = uFromDataList a b
        exact = min (exactP u) 1
        normal = normalP u
        normalUnc = normalUncorrectedP u
    return $ TestResult n m exact normal normalUnc</code></pre>
<p>Finally, we generate some random test results and print them to file.</p>
<pre><code>-- Generate some data for graphing and write to a file.
writeTestData :: Int -&gt; String -&gt; IO ()
writeTestData count fileName = withFile fileName WriteMode go where
  go handle = do
    rand &lt;- newStdGen
    let rnds rnd = rnd1 : rnds rnd2 where (rnd1, rnd2) = split rnd
        results = [(unGen genResult r n) | (r,n) &lt;- rnds rand `zip` [0..count]]
    mapM (hPutStrLn handle . show) results
    return ()</code></pre>
<p>For anyone who’s interested, the code is available <a href="https://gist.github.com/nc6/6509393">as a gist</a>.</p>

<div id="disqus">
    <div id="disqus_thread"></div>
    <script type="text/javascript">
        /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
        var disqus_shortname = 'categoricalimpredicative'; // Required - Replace example with your forum shortname

        /* * * DON'T EDIT BELOW THIS LINE * * */
        (function() {
            var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
            dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
            (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
        })();
    </script>
    <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
    <a href="http://disqus.com" class="dsq-brlink">blog comments powered by <span class="logo-disqus">Disqus</span></a>
</div>
        </div>
        <div id="footer">
            <div style="float:left;">
                Site proudly generated by
                <a href="http://jaspervdj.be/hakyll">Hakyll</a>
            </div>
            <div>
                All work on this site is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by/3.0/">Creative Commons Attribution 3.0 Unported License</a>.
            </div>
        </div>
    </body>
</html>
